File: mcp-server/internal/domains/mcpservers/delete.go
package mcpservers

import (
	"context"
	"fmt"

	mcpserverv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/mcpserver/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	"google.golang.org/grpc"
)

// Delete removes an MCP server identified by org and slug. It first resolves
// the org/slug pair to a resource ID via GetByReference, then calls the
// CommandController.Delete RPC with an ApiResourceDeleteInput. Both calls
// share a single gRPC connection.
func Delete(ctx context.Context, serverAddress, org, slug string) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			desc := fmt.Sprintf("MCP server %q in org %q", slug, org)

			queryClient := mcpserverv1.NewMcpServerQueryControllerClient(conn)
			mcpServer, err := queryClient.GetByReference(ctx, &apiresource.ApiResourceReference{
				Org:  org,
				Kind: apiresourcekind.ApiResourceKind_mcp_server,
				Slug: slug,
			})
			if err != nil {
				return "", domains.RPCError(err, desc)
			}

			cmdClient := mcpserverv1.NewMcpServerCommandControllerClient(conn)
			deleted, err := cmdClient.Delete(ctx, &apiresource.ApiResourceDeleteInput{
				ResourceId: mcpServer.GetMetadata().GetId(),
			})
			if err != nil {
				return "", domains.RPCError(err, desc)
			}

			return domains.MarshalJSON(deleted)
		})
}

File: mcp-server/src/domains/mcpservers/delete.ts
// MCP server delete path: resolve org/slug → id, then delete, both over a single
// shared transport.
// Go parity: mcp-server/internal/domains/mcpservers/delete.go.
//
// Outlier: unlike the agent/skill/workflow command controllers (which take a
// typed {X}Id), McpServerCommandController.delete takes the generic
// ApiResourceDeleteInput{resource_id}.

import { createClient } from "@connectrpc/connect";
import { McpServerSchema } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/api_pb";
import { McpServerCommandController } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/command_pb";
import { McpServerQueryController } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/query_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { withTransport } from "../client.js";
import { toProtoJson } from "../marshal.js";
import { rpcError } from "../rpcerr.js";

/**
 * Delete an MCP server by org and slug, returning the deleted MCP server as
 * protojson.
 */
export async function deleteMcpServer(
  serverAddress: string,
  token: string,
  org: string,
  slug: string,
): Promise<string> {
  const desc = `MCP server "${slug}" in org "${org}"`;
  return withTransport(serverAddress, token, async (transport, callOptions) => {
    const query = createClient(McpServerQueryController, transport);
    let id: string;
    try {
      const mcpServer = await query.getByReference(
        { org, kind: ApiResourceKind.mcp_server, slug },
        callOptions,
      );
      id = mcpServer.metadata?.id ?? "";
    } catch (err) {
      throw rpcError(err, desc);
    }

    const command = createClient(McpServerCommandController, transport);
    try {
      const deleted = await command.delete({ resourceId: id }, callOptions);
      return toProtoJson(McpServerSchema, deleted);
    } catch (err) {
      throw rpcError(err, desc);
    }
  });
}
